pkg/watermark/fetch/source_fetcher.go

```diff
@@ -67,8 +67,8 @@ func (e *sourceFetcher) GetWatermark(_ isb.Offset) wmb.Watermark {
 		if !p.IsActive() {
 			continue
 		}
-		if p.GetOffsetTimeline().GetHeadWatermark() < epoch {
-			epoch = p.GetOffsetTimeline().GetHeadWatermark()
+		if p.GetOffsetTimelines()[fromPartitionIdx].GetHeadWatermark() < epoch {
```

Comment:
please add nil check for array out of bound exception.

Reply:
that will never happen.

Reply:
Since these methods are only used within the platform, we can avoid these unnecessary if checks, since we control the input parameters. It makes sense to have those checks if the methods are exposed to users. And also if we are getting an invalid partitionCount it is better to catch those instead of ignoring them using an if check otherwise we will have a bigger problem.

Reply:
I don't fully agree with `Since these methods are only used within the platform, we can avoid these unnecessary if checks, since we control the input parameters.` I think each class/struct should take the responsibility of ensuring no NullPointer instead of the entire project/platform, because otherwise there could be very few cases when we need to do null check.

I agree with the second point that we'd rather to naturally panic instead of a if check. Resolving this comment.